pennylane/gradients/general_shift_rules.py

```diff
@@ -0,0 +1,97 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Contains a function for generating generalized parameter shift rules."""
+import numpy as np
+
+
+def get_shift_rule(frequencies, diff_shifts=None):

```

Comment:
I think decorating with `@functools.lru_cache()` makes sense for the reasons we've discussed before: a circuit ansätze with multiple layers of `CommutingEvolution` with isospectral generators makes multiple (potentially costly) calls to `get_shift_rule` with the same input arguments. I've included the decorator in the latest commit. This change also requires that the arguments to `get_shift_rule` are hashable, so I've changed `frequencies` and `shifts` to be tuples instead of lists in the docstring, examples, and tests.